utils/network.py

```diff
@@ -0,0 +1,255 @@
+        edge_trace["x"] += (
+            pos[source][0],
+            pos[target][0],
+            None,
+        )  # None creates a gap between line segments
+        edge_trace["y"] += (pos[source][1], pos[target][1], None)
+
+    edge_trace["hovertext"] = hovertext
+
+    # Define arrow symbol for edges
+    edge_trace["marker"] = dict(
+        symbol="arrow", color="#888", size=10, angleref="previous"
+    )
+
+    node_trace = go.Scatter(
+        x=[],
+        y=[],
+        text=[],
+        mode="markers",
+        hoverinfo="text",
+        marker=dict(showscale=True, colorscale="YlGnBu", size=10),
+    )
+    node_trace["marker"]["color"] = []
+
+    for node in G.nodes():
+        node_info = f"Name: {node}<br>"
+        for key, value in G.nodes[node].items():
+            node_info += f"{key}: {value}<br>"
+        node_trace["text"] += tuple([node_info])
+        # Get the classification value for the node
+        classification = G.nodes[node].get("classification", "neutral")
+        # Assign a color based on the classification value
+        if classification == "c":
+            color = "blue"
+        elif classification == "f":
+            color = "red"
+        else:
+            color = "green"  # Default color for unknown classification
+        node_trace["marker"]["color"] += tuple([color])
+
+        # Add node positions to the trace
+        node_trace["x"] += tuple([pos[node][0]])
+        node_trace["y"] += tuple([pos[node][1]])
+
+    # Define layout settings
+    layout = go.Layout(
+        title="Network Graph Indicating Campaign Contributions from 2018-2022",
+        titlefont=dict(size=16),
+        showlegend=True,
+        hovermode="closest",
+        margin=dict(b=20, l=5, r=5, t=40),
+        xaxis=dict(showgrid=True, zeroline=True, showticklabels=False),
+        yaxis=dict(showgrid=True, zeroline=True, showticklabels=False),
+    )
+
+    fig = go.Figure(data=[edge_trace, node_trace], layout=layout)
+    fig.show()
+
+
+# create pipeline
+
+
+def construct_network_graph(
+    start_year: int, end_year: int, dfs: list[pd.DataFrame]
+):
+    """Runs the network construction pipeline starting from 3 dataframes
+
+    Args:
+        start_year & end_year: the range of the desired data
+
+    Returns:
+    """
+    inds_df, orgs_df, transactions_df = dfs
+    transactions_df = transactions_df.loc[
+        (transactions_df.year >= start_year)
+        & (transactions_df.year <= end_year)
+    ]
+
+    aggreg_df = combine_datasets_for_network_graph(
+        [inds_df, orgs_df, transactions_df]
+    )
+    G = create_network_graph(aggreg_df)
+    plot_network_graph(G)
+    nx.write_adjlist(G, "Network Graph Node Data")
+
+
+def main():
+    """"""
+    text = input(
+        "Provide a range of desired years to extract data. Format is year1, \
+        year2. Ex: 2018, 2023"
+    )
+
+    assert len(text == 2)
+    start_year, end_year = text.split(",")
+    construct_network_graph(
```

Comment:
I would imagine this doesn't work because you are missing the dfs argument

Reply:
@averyschoen this is where I was stuck, bc these would come from the record_linkage pipeline. How should I approach this?

Reply:
work with adil here -- it look's like he could just call construct_network_graph() in his main function of linkage_pipeline.py